tests/testInitialSetupConsent.php

```diff
@@ -0,0 +1,291 @@
+
+		$redirect_url = parse_url( $caught_redirect['location'] );
+
+		$this->assertEquals( '/wp-admin/admin.php', $redirect_url['path'] );
+		$this->assertContains( 'page=wpa0', $redirect_url['query'] );
+		$this->assertContains( 'error=cant_create_client', $redirect_url['query'] );
+
+		$this->assertCount( 1, self::$error_log->get() );
+	}
+
+	/**
+	 * Test that an existing DB connection is used instead of creating one.
+	 */
+	public function testThatExistingConnectionSkipsCreation() {
+		$this->startHttpMocking();
+		$this->startRedirectHalting();
+
+		self::$opts->set( 'client_signing_algorithm', 'HS256' );
+
+		$setup_consent = new WP_Auth0_InitialSetup_Consent( self::$opts );
+		$test_token    = implode( '.', [ uniqid(), uniqid(), uniqid() ] );
+
+		// Mock consecutive HTTP calls.
+		$this->http_request_type = [
+			// Successful client creation.
+			'success_create_client',
+			// Found a connection with the same name.
+			'success_get_existing_conn',
+			// Client grant created successfully.
+			'success_create_empty_body',
+		];
+
+		$caught_redirect = [];
+		try {
+			$setup_consent->callback_with_token( 'test-wp.auth0.com', $test_token, 'social' );
+		} catch ( Exception $e ) {
+			$caught_redirect = unserialize( $e->getMessage() );
+		}
+
+		$this->assertNotEmpty( $caught_redirect );
+		$this->assertEquals( 302, $caught_redirect['status'] );
+
+		$redirect_url = parse_url( $caught_redirect['location'] );
+
+		$this->assertEquals( '/wp-admin/admin.php', $redirect_url['path'] );
+		$this->assertContains( 'page=wpa0-setup', $redirect_url['query'] );
+		$this->assertContains( 'step=2', $redirect_url['query'] );
+		$this->assertContains( 'profile=social', $redirect_url['query'] );
+
+		$this->assertEquals( 'TEST_CLIENT_ID', self::$opts->get( 'client_id' ) );
+		$this->assertEquals( 'TEST_CLIENT_SECRET', self::$opts->get( 'client_secret' ) );
+		$this->assertEquals( 1, self::$opts->get( 'db_connection_enabled' ) );
+		$this->assertEquals( 'TEST_CONN_ID', self::$opts->get( 'db_connection_id' ) );
+		$this->assertEquals( 'good', self::$opts->get( 'password_policy' ) );
+
+		$this->assertEmpty( self::$error_log->get() );
+	}
+
+	/**
+	 * Test that an connection is created and the Client Grant fails.
+	 */
+	public function testThatNewConnectionIsCreatedAndFailedClientGrantRedirects() {
+		$this->startHttpMocking();
+		$this->startRedirectHalting();
+
+		$setup_consent = new WP_Auth0_InitialSetup_Consent( self::$opts );
+		$test_token    = implode( '.', [ uniqid(), uniqid(), uniqid() ] );
+
+		self::$opts->set( 'client_signing_algorithm', 'HS256' );
+
+		// Mock consecutive HTTP calls.
+		$this->http_request_type = [
+			// Successful client creation.
+			'success_create_client',
+			// Get en existing connection enabled for this client.
+			'success_get_connections',
+			// Connection updated successfully.
+			'success_update_connection',
+			// Connection created successfully.
+			'success_create_connection',
+			// Client grant failed.
```

Comment:
why would this happen? Seems like this should belong in a different test case than "create new connection succeeds"

Reply:
Maybe the token doesn't have the right scopes or HTTP error or any number of things.

Reply:
In that case, that could be moved to a different test other than the "happy path" of creating the connection when missing